src/redis_cmd.cc

```diff
@@ -3270,57 +3270,72 @@ class CommandPublish : public Commander {
+  output->append(sub_name.empty() ? Redis::NilString() : Redis::BulkString(sub_name));
+  output->append(Redis::Integer(num));
+}
+
 class CommandSubscribe : public Commander {
  public:
-  CommandSubscribe() : Commander("subcribe", -2, false) {}
+  CommandSubscribe() : Commander("subscribe", -2, false) {}
   Status Execute(Server *svr, Connection *conn, std::string *output) override {
     for (unsigned i = 1; i < args_.size(); i++) {
       conn->SubscribeChannel(args_[i]);
-      output->append(Redis::MultiLen(3));
-      output->append(Redis::BulkString("subscribe"));
-      output->append(Redis::BulkString(args_[i]));
-      output->append(Redis::Integer(conn->SubscriptionsCount()));
+      SubscribeCommmandReply(output, "subscribe", args_[i],
+        conn->SubscriptionsCount() + conn->PSubscriptionsCount());
     }
     return Status::OK();
   }
 };
 
 class CommandUnSubscribe : public Commander {
  public:
-  CommandUnSubscribe() : Commander("unsubcribe", -1, false) {}
+  CommandUnSubscribe() : Commander("unsubscribe", -1, false) {}
   Status Execute(Server *svr, Connection *conn, std::string *output) override {
-    if (args_.size() > 1) {
-      conn->UnSubscribeChannel(args_[1]);
-    } else {
+    if (args_.size() == 1) {
       conn->UnSubscribeAll();
+      SubscribeCommmandReply(output, "unsubscribe", "",
+        conn->SubscriptionsCount() + conn->PSubscriptionsCount());
+    } else {
+      for (unsigned i = 1; i < args_.size(); i++) {
+        conn->UnSubscribeChannel(args_[i]);
+        SubscribeCommmandReply(output, "unsubscribe", args_[i],
+          conn->SubscriptionsCount() + conn->PSubscriptionsCount());
+      }
     }
     return Status::OK();
   }
 };
 
 class CommandPSubscribe : public Commander {
  public:
-  CommandPSubscribe() : Commander("psubcribe", -2, false) {}
+  CommandPSubscribe() : Commander("psubscribe", -2, false) {}
   Status Execute(Server *svr, Connection *conn, std::string *output) override {
     for (unsigned i = 1; i < args_.size(); i++) {
       conn->PSubscribeChannel(args_[i]);
-      output->append(Redis::MultiLen(3));
-      output->append(Redis::BulkString("psubscribe"));
-      output->append(Redis::BulkString(args_[i]));
-      output->append(Redis::Integer(conn->PSubscriptionsCount()));
+      SubscribeCommmandReply(output, "psubscribe", args_[i],
+        conn->SubscriptionsCount() + conn->PSubscriptionsCount());
     }
     return Status::OK();
   }
 };
 
 class CommandPUnSubscribe : public Commander {
  public:
-  CommandPUnSubscribe() : Commander("punsubcribe", -1, false) {}
+  CommandPUnSubscribe() : Commander("punsubscribe", -1, false) {}
   Status Execute(Server *svr, Connection *conn, std::string *output) override {
-    if (args_.size() > 1) {
-      conn->PUnSubscribeChannel(args_[1]);
+    if (args_.size() == 1) {
+      conn->PUnSubscribeAll();;
+      SubscribeCommmandReply(output, "punsubscribe", "",
+        conn->SubscriptionsCount() + conn->PSubscriptionsCount());
```

Comment:
in punsubscribe all, a message for every punsubscribed channel_patten will be sent to the client in stock redis (verified on redis v5.0.4)， is this response on purpose？

Reply:
You are right, i miss that, please have a look again.